app/controllers/incomeSources/add/IncomeSourceCheckDetailsController.scala

```diff
@@ -120,25 +123,27 @@ class IncomeSourceCheckDetailsController @Inject()(val checkDetailsView: IncomeS
     }
   }
 
-  private def getPropertyModel(incomeSourceType: IncomeSourceType)(implicit user: MtdItUser[_]): Future[Either[Throwable, CheckPropertyViewModel]] = {
-    sessionService.getMongoKeyTyped[LocalDate](dateStartedField, JourneyType(Add, incomeSourceType)).flatMap { startDate: Either[Throwable, Option[LocalDate]] =>
-      sessionService.getMongoKeyTyped[String](incomeSourcesAccountingMethodField, JourneyType(Add, incomeSourceType)).map { accMethod: Either[Throwable, Option[String]] =>
-        (startDate, accMethod) match {
-          case (Right(dateMaybe), Right(methodMaybe)) =>
-            (dateMaybe, methodMaybe) match {
-              case (Some(date), Some(method)) =>
-                Right(CheckPropertyViewModel(
-                  tradingStartDate = date,
-                  cashOrAccrualsFlag = method,
-                  incomeSourceType = incomeSourceType
-                ))
-              case (_, _) =>
-                Left(new Error(s"Start date or accounting method not found in session. Start date: $dateMaybe, AccMethod: $methodMaybe"))
-            }
-          case (_, _) => Left(new Error(s"Error while retrieving date started or accounting method from session"))
+  private def getPropertyModel(incomeSourceType: IncomeSourceType)
+                              (implicit user: MtdItUser[_]): Future[Either[Throwable, CheckPropertyViewModel]] = {
+    sessionService.getMongoKeyTyped[LocalDate](dateStartedField, JourneyType(Add, incomeSourceType))
+      .flatMap { startDate: Either[Throwable, Option[LocalDate]] =>
+        sessionService.getMongoKeyTyped[String](incomeSourcesAccountingMethodField, JourneyType(Add, incomeSourceType)).map { accMethod: Either[Throwable, Option[String]] =>
+          (startDate, accMethod) match {
+            case (Right(dateMaybe), Right(methodMaybe)) =>
+              (dateMaybe, methodMaybe) match {
+                case (Some(date), Some(method)) =>
+                  Right(CheckPropertyViewModel(
+                    tradingStartDate = date,
+                    cashOrAccrualsFlag = method,
+                    incomeSourceType = incomeSourceType
+                  ))
+                case (_, _) =>
+                  Left(new Error(s"Start date or accounting method not found in session. Start date: $dateMaybe, AccMethod: $methodMaybe"))
+              }
+            case (_, _) => Left(new Error(s"Error while retrieving date started or accounting method from session"))
+          }
```

Comment:
I thought about adding another constructor for Option[IncomeSourceId] like so:

`object IncomeSourceId {
  def mkIncomeSourceId(incomeSourceAsString: String): IncomeSourceId = {
    new IncomeSourceId(incomeSourceAsString)
  }

  def mkIncomeSourceIdFromOption(incomeSourceAsStringMaybe: Option[String]): Either[Throwable, Option[IncomeSourceId]] = {
    incomeSourceAsStringMaybe match {
      case Some(incomeSourceIdAsString: String) => Right(Option(new IncomeSourceId(incomeSourceIdAsString)))
      case None => Left(new Error(s"Error while creating incomeSourceId object. incomeSourceAsStringMaybe was empty"))
    }
  }
}`

What do you think?

Reply:
this method is redundant, you call always do something like:

```
val idMaybe: Option[String] = Some("String")
val incomeSourceIdMaybe = idMaybe.map(mkIncomeSourceId)
```